zebra/zebra_dplane.c

```diff
@@ -7328,18 +7328,46 @@ static void dplane_thread_loop(struct event *event)
 	/* Locate initial registered provider */
 	prov = dplane_prov_list_first(&zdplane_info.dg_providers);
 
-	/* Move new work from incoming list to temp list */
-	for (counter = 0; counter < limit; counter++) {
-		ctx = dplane_ctx_list_pop(&zdplane_info.dg_update_list);
-		if (ctx) {
-			ctx->zd_provider = prov->dp_id;
+	curr = atomic_load_explicit(&prov->dp_in_queued, memory_order_relaxed);
+	out_curr = atomic_load_explicit(&prov->dp_out_queued,
+					memory_order_relaxed);
 
-			dplane_ctx_list_add_tail(&work_list, ctx);
-		} else {
-			break;
+	if (curr >= (uint64_t)limit) {
+		reschedule = true;
+		if (IS_ZEBRA_DEBUG_DPLANE_DETAIL)
+			zlog_debug("%s: Current first provider(%s) Input queue is %" PRIu64
+				   ", holding off work",
+				   __func__, prov->dp_name, curr);
+		counter = 0;
+	} else if (out_curr >= (uint64_t)limit) {
```

Comment:
Can it be a situation where the input and output queue is hitting the limit?

Reply:
Yeah but it doesn't matter from my perspective.  Both cases we do the same thing.  It's sufficient to just state the first one encountered in a debug and move on.